src/jit/gentree.cpp

```diff
@@ -611,6 +611,60 @@ void Compiler::fgWalkAllTreesPre(fgWalkPreFn* visitor, void* pCallBackData)
 }
 
 // ------------------------------------------------------------------------------------------
+// gtCheckReg: Checks that the register has either been set, or is still set to the "no register
+// assignment" value, depending upon whether it fits any of the special cases for multi-reg call nodes.
+//
+// Arguments:
+//     compiler  -  compiler instance
+//
+// Return Value:
+//     None
+//
+// Notes:
+//     This also sets the gtRegTag to GT_REGTAG_REG.
+//
+void GenTree::gtCheckReg()
+{
+#ifdef DEBUG
+    assert((gtRegTag != GT_REGTAG_NONE) || (gtRegNum == REG_NA));
+#if CPU_LONG_USES_REGPAIR
+    assert(!"gtSetRegDefault not supported for REGPAIR");
+#endif // CPU_LONG_USES_REGPAIR
+    if (gtRegTag == GT_REGTAG_NONE)
+    {
+        gtRegTag = GT_REGTAG_REG;
+
```

Comment:
I'm not really a fan of `gtRegTag`, and this latest round of changes makes me less so.
The problem is that, during `Lowering`, we need to record some register requirements (e.g. for call arguments), and the reasonable way to do that is on the `GenTree` itself. Previously, this was being done by the hack of having `TreeNodeInfo::Initialize()` set the candidates based on the value of `gtRegNum` before calling `gtClearReg()`. Then the register would get restored to its intended value during register allocation.
Since I want to eliminate `gtLsraInfo` I can't use that to record this information (not that it was a great approach to begin with, IMHO), but since these are being set during `Lowering` and we are adding new nodes, we either have to call `gtClearReg()` on each node before `Lowering` and then make sure we call it on each and every new node we add, OR explicitly clear the `gtRegTag`. I went down the former path, and it required touching nodes that `Lowering` wouldn't otherwise need to mess with.